scripts/prune_lm_dir.py

```diff
@@ -328,6 +351,160 @@ def FinalizeOutput(final_work_out):
         os.remove(args.lm_dir_out + "/num_splits")
 
 
+def MatchTargetSize(num_ngrams):
+    assert(args.target_num_ngrams > 0)
+    return abs(num_ngrams - args.target_num_ngrams) / float(args.target_num_ngrams) < args.tolerance
+
+def IterateOnce(threshold, step, iter, recovery_step, recovery_size):
+    """Prune with a specific threshold and check whether the resulting LM matches the target-num-ngrams
+
+    Args:
+        threshold: the threshold used to prune LM
+        step: the current step-index (each individual operation, such as pruning or E-M, counts as one step).
+        iter: index of current iteration. We only increase the iteration index when we do an actual pruning step,
+            so it will be <= the step index.'
+        recovery_step, recovery_size: the step-index and num-grams
+            for LM that is most recent and smallest but larger than target-num-ngrams.
+            This is used as the starting point for the iteration done by this function.
+    Returns:
+        A tuple with the first element is a bool variable indicating whether we already got the right threshold.
+        And the remaining are updated value for (step, iter, recovery_step, recovery_size)
+    Example Usage:
+
+        step = 0
+        recovery_step = step
+        recovery_size = 0
+        iter = 0
+
+        threshold = init_threshold()
+        while True:
+            threshold = EstimateThreshold(args.target_num_ngrams)
+
+            (success, step, iter, recovery_step, recovery_size) = \
+                    IterateOnce(threshold, step, iter, recovery_step, recovery_size)
+            if success:
+                break
+
+    """
+    global logprob_changes, steps, final_num_ngrams
+
+    if step > 0:
+        scale = threshold / args.threshold
+        steps += 'prune*{0} EM EM EM'.format(scale).split()
+
+    # Prune step
+    logprob_changes.append(RunStep(step, in_step=recovery_step))
+    step += 1
+    while step < len(steps): # EM steps
+      logprob_changes.append(RunStep(step))
+      step += 1
+    iter += 1
+
+    if MatchTargetSize(final_num_ngrams):
+        return (True, step, iter, recovery_step, recovery_size)
+
+    if iter > args.max_iter:
+        sys.exit("prune_lm_dir.py: Too many iterations, please set a higher threshold and rerun.")
+
+    # always prune from LM larger than target_num_ngrams
+    if final_num_ngrams > args.target_num_ngrams and (recovery_size <= 0 or final_num_ngrams < recovery_size):
```

Comment:
Because you are now using final_num_ngrams as the num-ngrams from the most recent pruning step, please rename the variable to current_num_ngrams.